crates/net/eth-wire/src/p2pstream.rs

```diff
@@ -606,32 +565,40 @@ where
         // all messages sent in this stream are subprotocol messages, so we need to switch the
         // message id based on the offset
         compressed[0] = item[0] + MAX_RESERVED_MESSAGE_ID + 1;
-        this.outgoing_messages.push_back(compressed.freeze());
+        this.inner.start_send_unpin(compressed.freeze()).map_err(P2PStreamError::from)?;
 
         Ok(())
     }
 
     /// Returns `Poll::Ready(Ok(()))` when no buffered items remain.
     fn poll_flush(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Result<(), Self::Error>> {
         let mut this = self.project();
-        let poll_res = loop {
-            match this.inner.as_mut().poll_ready(cx) {
-                Poll::Pending => break Poll::Pending,
-                Poll::Ready(Err(err)) => break Poll::Ready(Err(err.into())),
+
+        // Check if we need to send a disconnect message or a ping message, if so, send them
+        match (this.pending_disconnect.take(), this.pending_ping.take()) {
+            (Some(disconnect), _ping) => match this.inner.poll_ready_unpin(cx) {
                 Poll::Ready(Ok(())) => {
-                    let Some(message) = this.outgoing_messages.pop_front() else {
-                        break Poll::Ready(Ok(()))
-                    };
-                    if let Err(err) = this.inner.as_mut().start_send(message) {
-                        break Poll::Ready(Err(err.into()))
-                    }
+                    this.inner.start_send_unpin(disconnect)?;
                 }
-            }
-        };
-
-        ready!(this.inner.as_mut().poll_flush(cx))?;
-
-        poll_res
+                Poll::Ready(Err(e)) => return Poll::Ready(Err(e.into())),
+                Poll::Pending => {
+                    *this.pending_disconnect = Some(disconnect);
+                    return Poll::Pending;
+                }
+            },
+            (None, Some(ping)) => match this.inner.poll_ready_unpin(cx) {
+                Poll::Ready(Ok(())) => {
+                    this.inner.start_send_unpin(ping)?;
+                }
+                Poll::Ready(Err(e)) => return Poll::Ready(Err(e.into())),
+                Poll::Pending => {
+                    *this.pending_ping = Some(ping);
+                    return Poll::Pending;
+                }
+            },
+            (None, None) => (), // Nothing pending to send
+        }
+        this.inner.poll_flush_unpin(cx).map_err(P2PStreamError::from)
```

Comment:
I think having only one `pending_message` would simplify and make cleaner, but at the same time, it could make the disconnect flow harder to understand? lmk what you think haha